File: python/src/dsa/data_structures/singly_linked_list.py
"""Singly linked list implementation.

Time: O(1) for push_front, O(n) for push_back, search
Space: O(n)
"""

from typing import Generic, TypeVar, Optional

T = TypeVar("T")


class _Node(Generic[T]):
    """Internal node class."""

    def __init__(self, value: T) -> None:
        self.value = value
        self.next: Optional[_Node[T]] = None


class SinglyLinkedList(Generic[T]):
    """A generic singly linked list implementation."""

    def __init__(self) -> None:
        self._head: Optional[_Node[T]] = None
        self._len = 0

    def push_front(self, value: T) -> None:
        """Add a value to the front of the list."""
        node = _Node(value)
        node.next = self._head
        self._head = node
        self._len += 1

    def pop_front(self) -> Optional[T]:
        """Remove and return the front value, or None if empty."""
        if self._head is None:
            return None
        value = self._head.value
        self._head = self._head.next
        self._len -= 1
        return value

    def push_back(self, value: T) -> None:
        """Add a value to the back of the list."""
        node = _Node(value)
        if self._head is None:
            self._head = node
        else:
            current = self._head
            while current.next is not None:
                current = current.next
            current.next = node
        self._len += 1

    def peek_front(self) -> Optional[T]:
        """Return the front value without removing it, or None if empty."""
        if self._head is None:
            return None
        return self._head.value

    def contains(self, value: T) -> bool:
        """Return True if the value is in the list."""
        current = self._head
        while current is not None:
            if current.value == value:
                return True
            current = current.next
        return False

    def is_empty(self) -> bool:
        """Return True if the list is empty."""
        return self._head is None

    def __len__(self) -> int:
        """Return the number of elements."""
        return self._len

    def clear(self) -> None:
        """Remove all elements."""
        self._head = None
        self._len = 0

File: nim/src/dsa/data_structures/singly_linked_list.nim
## Singly Linked List - Forward-only linked list
##
## Time Complexity:
## - prepend: O(1)
## - append: O(n) or O(1) with tail pointer
## - get: O(n)
## - removeAt: O(n)
## - find: O(n)

type
  Node[T] = ref object
    data: T
    next: Node[T]

  SinglyLinkedList*[T] = object
    head: Node[T]
    tail: Node[T]
    count: int

proc newSinglyLinkedList*[T](): SinglyLinkedList[T] =
  ## Create a new empty singly linked list
  SinglyLinkedList[T](head: nil, tail: nil, count: 0)

proc prepend*[T](list: var SinglyLinkedList[T], item: T) =
  ## Add an item to the front
  let node = Node[T](data: item, next: list.head)
  list.head = node
  if list.tail == nil:
    list.tail = node
  inc list.count

proc append*[T](list: var SinglyLinkedList[T], item: T) =
  ## Add an item to the back
  let node = Node[T](data: item, next: nil)
  if list.tail == nil:
    list.head = node
    list.tail = node
  else:
    list.tail.next = node
    list.tail = node
  inc list.count

proc get*[T](list: SinglyLinkedList[T], index: int): T =
  ## Get item at index
  ## Raises IndexDefect if index is out of bounds
  if index < 0 or index >= list.count:
    raise newException(IndexDefect, "Index out of bounds")
  var current = list.head
  for _ in 0 ..< index:
    current = current.next
  current.data

proc `[]`*[T](list: SinglyLinkedList[T], index: int): T =
  ## Get item at index using [] operator
  list.get(index)

proc set*[T](list: var SinglyLinkedList[T], index: int, item: T) =
  ## Set item at index
  ## Raises IndexDefect if index is out of bounds
  if index < 0 or index >= list.count:
    raise newException(IndexDefect, "Index out of bounds")
  var current = list.head
  for _ in 0 ..< index:
    current = current.next
  current.data = item

proc removeFirst*[T](list: var SinglyLinkedList[T]): T =
  ## Remove and return the first item
  ## Raises IndexDefect if list is empty
  if list.head == nil:
    raise newException(IndexDefect, "List is empty")
  result = list.head.data
  list.head = list.head.next
  if list.head == nil:
    list.tail = nil
  dec list.count

proc removeAt*[T](list: var SinglyLinkedList[T], index: int): T =
  ## Remove and return item at index
  ## Raises IndexDefect if index is out of bounds
  if index < 0 or index >= list.count:
    raise newException(IndexDefect, "Index out of bounds")
  if index == 0:
    return list.removeFirst()
  var current = list.head
  for _ in 0 ..< index - 1:
    current = current.next
  result = current.next.data
  current.next = current.next.next
  if current.next == nil:
    list.tail = current
  dec list.count

proc find*[T](list: SinglyLinkedList[T], item: T): int =
  ## Find index of item, returns -1 if not found
  var current = list.head
  var index = 0
  while current != nil:
    if current.data == item:
      return index
    current = current.next
    inc index
  -1

proc contains*[T](list: SinglyLinkedList[T], item: T): bool =
  ## Check if list contains item
  list.find(item) != -1

proc isEmpty*[T](list: SinglyLinkedList[T]): bool =
  ## Check if the list is empty
  list.count == 0

proc size*[T](list: SinglyLinkedList[T]): int =
  ## Return the number of items
  list.count

proc len*[T](list: SinglyLinkedList[T]): int =
  ## Return the number of items (alias for size)
  list.count

proc clear*[T](list: var SinglyLinkedList[T]) =
  ## Remove all items
  list.head = nil
  list.tail = nil
  list.count = 0

proc toSeq*[T](list: SinglyLinkedList[T]): seq[T] =
  ## Convert to sequence
  result = @[]
  var current = list.head
  while current != nil:
    result.add(current.data)
    current = current.next

iterator items*[T](list: SinglyLinkedList[T]): T =
  ## Iterate over items
  var current = list.head
  while current != nil:
    yield current.data
    current = current.next

when isMainModule:
  var list = newSinglyLinkedList[int]()
  list.append(1)
  list.append(2)
  list.prepend(0)
  assert list[0] == 0
  assert list[1] == 1
  assert list[2] == 2
  assert list.size() == 3
  assert list.removeFirst() == 0
  assert list.size() == 2
  assert list.find(2) == 1
  assert list.contains(1)
  echo "SinglyLinkedList tests passed!"
